nexus/reconfigurator/simulation/src/state.rs

```diff
@@ -0,0 +1,510 @@
+                        blueprint_id,
+                    ));
+                }
+            }
+        }
+
+        self.system.description_mut().service_ip_pool_ranges(
+            state.planning_input.service_ip_pool_ranges().to_vec(),
+        );
+        res.notices.push(format!(
+            // TODO: better output format?
+            "loaded service IP pool ranges: {:?}",
+            state.planning_input.service_ip_pool_ranges()
+        ));
+
+        self.system.set_internal_dns(state.internal_dns);
+        self.system.set_external_dns(state.external_dns);
+
+        let nnames = state.external_dns_zone_names.len();
+        if nnames > 0 {
+            if nnames > 1 {
+                res.warnings.push(format!(
+                    "found {} external DNS names; using only the first one",
+                    nnames
+                ));
+            }
+            self.config.set_external_dns_zone_name(
+                state.external_dns_zone_names[0].clone(),
+            );
+        }
+
+        // TODO: Currently this doesn't return notices for DNS and silo names.
+        // The only caller of this function prints them separately after
+        // committing this state. We may want to record this information in the
+        // MergeResult instead.
+
+        // TODO: log what happened here. This is a cross-cutting change so we
+        // may want to log it as a single big entry (like
+        // SimSystemBuilder::load_example) rather than lots of little ones.
+    }
+
+    /// Commit the current state to the store, returning the new state's UUID.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `sim` is not the same simulator that created this state. This
+    /// should ordinarily never happen and always indicates a programming
+    /// error.
+    #[must_use = "callers should update their pointers with the returned UUID"]
+    pub fn commit(
+        self,
+        description: String,
+        sim: &mut Simulator,
+    ) -> ReconfiguratorSimUuid {
+        // Check for unrelated histories.
+        if !std::ptr::eq(sim.root_state(), self.root_state.inner()) {
+            panic!(
+                "this state was created by a different simulator than the one \
+                 it is being committed to"
+            );
+        }
+
+        let id = sim.next_sim_uuid();
+        let (system, system_log) = self.system.into_parts();
+        let (config, config_log) = self.config.into_parts();
+        let (rng, rng_log) = self.rng.into_parts();
+        let log = SimStateLog {
+            system: system_log,
+            config: config_log,
+            rng: rng_log,
+        };
+        let state = SimState {
+            root_state: self.root_state,
+            id,
+            description,
+            parent: Some(self.parent),
+            generation: self.parent_gen.next(),
+            system,
+            config,
+            rng,
+            log,
+        };
+        sim.add_state(Arc::new(state));
```

Comment:
There are a bunch of comments around why things look this way, but I'm still not sold that it's the right way to go. For example, there's a requirement that `sim.next_sim_uuid()` is called before `sim.add_state()`, but this cannot be checked at compile time. I think rather than passing in a `Simulator` to this method, the `Simulator` should have a method to return a mutable reference to the "next" `SimStateBuilder`, or even better, take a closure that can generate the next `SimState` when passed an `&SimStateBuilder`. Then when that closure returns the  committed state can be added by the simulator.

In short, I'd like this all to be "driven" via the `Simulator` without having to pass in the simulator. 

It's totally possible I'm missing something here, but I think there may be room to simplify the API a bit and would like to discuss more.

Reply:
As we discussed, I would prefer to keep it this way because it means that if the caller decides to throw away an intermediate state, they can easily do so. This is the general model I've seen other tools follow, and while it's slightly looser at compile time I think the flexibility is worth it.

There are some dependencies that can't easily be checked at compile-time, but only one that's externally visible: the simulator passed in must be the one the state originated from. I don't think this is a major burden.

> For example, there's a requirement that sim.next_sim_uuid() is called before sim.add_state(), but this cannot be checked at compile time.

This is true, but it's not externally visible -- this is managed purely within the reconfigurator-sim code.